File: 11. backTrace/[77]组合.java
//给定两个整数 n 和 k，返回 1 ... n 中所有可能的 k 个数的组合。 
//
// 示例: 
//
// 输入: n = 4, k = 2
//输出:
//[
//  [2,4],
//  [3,4],
//  [2,3],
//  [1,2],
//  [1,3],
//  [1,4],
//] 
// Related Topics 回溯算法 
// 👍 580 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    private void backTrace(int begin, int end, int k, List<Integer> box, List<List<Integer>> ans) {
        if (box.size() == k) {
            ans.add(new ArrayList<>());
            for (Integer x : box) {
                ans.get(ans.size()-1).add(x);
            }
        }

        if (begin >= end || box.size() > k) {
            return;
        }

        for (int i = begin; i < end; i++) {
            box.add(i);
            backTrace(i+1, end, k, box, ans);
            box.remove(box.size()-1);
        }
    }
    public List<List<Integer>> combine(int n, int k) {
        List<Integer> box = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        backTrace(1, n+1, k, box, ans);
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 11. backTrace/[77]组合.py
# 给定两个整数 n 和 k，返回范围 [1, n] 中所有可能的 k 个数的组合。 
# 
#  你可以按 任何顺序 返回答案。 
# 
#  
# 
#  示例 1： 
# 
#  
# 输入：n = 4, k = 2
# 输出：
# [
#   [2,4],
#   [3,4],
#   [2,3],
#   [1,2],
#   [1,3],
#   [1,4],
# ] 
# 
#  示例 2： 
# 
#  
# 输入：n = 1, k = 1
# 输出：[[1]] 
# 
#  
# 
#  提示： 
# 
#  
#  1 <= n <= 20 
#  1 <= k <= n 
#  
# 
#  Related Topics 回溯 👍 1535 👎 0

from typing import List
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def backtrace(self, end, k, i, box, ans):
        if len(box)==k:
            temp = []
            for x in box:
                temp.append(x)
            ans.append(temp)

        if len(box)>=k or i>=end:
            return

        for j in range(i, end):
            box.append(j)
            self.backtrace(end, k, j+1, box, ans)
            box.pop()

    def combine(self, n: int, k: int) -> List[List[int]]:
        box = []
        ans = []
        self.backtrace(n+1, k, 1, box, ans)
        return ans
# leetcode submit region end(Prohibit modification and deletion)
